mqtt/mqtt-broker/src/persist.rs

```diff
@@ -272,22 +294,68 @@ impl From<ConsolidatedState> for BrokerSnapshot {
     }
 }
 
+/// Actual representation of session state data that is saved to disk.
 #[derive(Deserialize, Serialize)]
 struct ConsolidatedSession {
     client_info: ClientInfo,
     subscriptions: HashMap<String, Subscription>,
-    waiting_to_be_sent: Vec<SimplifiedPublication>,
+    waiting_to_be_sent: VecDeque<PublicationRef>,
+    waiting_to_be_acked: VecDeque<InFlightPublicationRef>,
     last_active: DateTime<Utc>,
 }
 
+/// Represents a queued publication
+/// that has a reference to a payload in `ConsolidatedState`
+/// but not yet in-flight.
+#[derive(Deserialize, Serialize)]
+struct PublicationRef {
+    topic_name: String,
+    qos: QoS,
+    retain: bool,
+    payload: u64,
+}
+
+/// Represents an in-flight publication (has packet id)
+/// and a reference to a payload in `ConsolidatedState`.
 #[derive(Deserialize, Serialize)]
-struct SimplifiedPublication {
+struct InFlightPublicationRef {
     topic_name: String,
-    qos: crate::proto::QoS,
+    packet_identifier_dup_qos: PacketIdentifierDupQoS,
     retain: bool,
     payload: u64,
 }
 
+#[allow(clippy::mutable_key_type)]
```

Comment:
is it still an issue?

Reply:
yes, it complains about `HashMap<Bytes, u64>` where it considers `Bytes` as mutable....